File: Rust/LinkedList/tstLinkedList.cpp
#include <iostream>
#include <exception>

#include "LinkedList.hpp"

using namespace std;

int main() {
  LinkedList<int> list;
  cout << list << endl;
  try {
	cout << list.front() << endl;
  } catch (std::exception& e) {
	cout << "Exception caught: " << e.what() << endl;
  }

  list.addFront('a');           // Should be invalid, but works
  list.addFront(21);
  list.addFront(14);
  list.addFront(7);

  cout << "Initial list:\n";
  cout << list << endl;


  // Check iterative reversing:
  cout << "\nAfter iterative reverse:\n";
  list.reverse();
  cout << list << endl;

  // Check recursive reversing:
  cout << "\nAfter recursive reverse:\n";
  list.reverseRecursive();
  cout << list << endl;

  // Check deleting nodes:
  cout << "\nAfter deleting non-existing key:\n";
  list.deleteKey(0);
  cout << list << endl;

  cout << "\nAfter deleting head key:\n";
  list.deleteKey(list.head()->value());
  cout << list << endl;

  cout << "\nAfter deleting middle key:\n";
  list.deleteKey(list.head()->next()->value());
  cout << list << endl;

  cout << "\nAfter deleting key:\n";
  list.deleteKey(list.head()->next()->value());
  cout << list << endl;

  cout << "\nCreate an Linked List Int:\n";
  Int num(100);
  cout << num << endl;
  num = Int(200);
  cout << num << endl;

  Int a(123);
  Int b(897);
  Int c = a + b;
  cout << a << "+" << b << "=" << c << endl;
  

  LinkedList<int> listArb;
  listArb.addFront(15);
  listArb.addFront(21);
  listArb.addFront(14);
  listArb.addFront(7);
  listArb.head()->setArb(listArb.head()->next()->next());
  listArb.head()->next()->next()->setArb(listArb.head());

  cout << "List with arbitraries for deep copy:"  << listArb << endl;
  listArb.printArb();
  LinkedList<int> listArbCopy = listArb.deepCopy();
  // listArbCopy.setHead(deep_copy_arbitrary_pointer(listArb.head()));

  // Change some stuff to check if the linked lists are really separate
  listArbCopy.head()->next()->setValue(123);
  listArb.head()->setValue(321);
  
  cout << listArbCopy << endl;
  cout << listArb << endl;
  
  // Check the sorting algos:
  LinkedList<int> sort;
  sort.addFront(11);
  sort.addFront(82);
  sort.addFront(23);
  sort.addFront(29);

  cout << sort << endl;
  sort.sort();
  cout << sort << endl;
  
  
  return 0;
}

File: Rust/LinkedList/tstLinkedList.java
import LinkedListJava.*;

public class tstLinkedList{
    public static void main(String[] args) {
        LinkedList<Integer> list = new LinkedList<>();
        try {
            list.front();
        } catch (Exception e) {
            System.out.println ("Exception:\n" + e);
        }
        // Node<Integer> head = new Node<Integer>();
        // list.setHead(head);
        
        System.out.println("Empty list:\n" + list);
        list.addFront(28);
        list.addFront(21);
        list.addFront(14);
        list.addFront(7);
        System.out.println("Initial list:\n" + list);
        list.removeFront();
        System.out.println("After removing front:\n" + list);
        list.addFront(7);

        list.reverse();
        System.out.println("After iterative reverse:\n" + list);

        list.reverseRecursive();
        System.out.println("After recursive reverse:\n" + list);

        list.deleteKey(0);
        System.out.println("After trying to remove non-existent element:\n" + list);

        list.deleteKey(14);
        System.out.println("After removing mid-element:\n" + list);

        list.deleteKey(7);
        System.out.println("After removing head element:\n" + list);

        list.deleteKey(28);
        System.out.println("After removing the last element:\n" + list);

        Int A = new Int(123);
        System.out.println("A:\n" + A);
        Int B = new Int(897);
        System.out.println("B:\n" + B);
        Int C = new Int();
        C.add(A, B);
        System.out.println("A + B:\n" + C);

        System.out.println("\nCreate a linked list with arbitrary pointers:\n");
        LinkedList<Integer> listArb = new LinkedList<>();
        listArb.addFront(21);
        listArb.addFront(14);
        listArb.addFront(7);
        listArb.head().setArb(listArb.head().next().next());
        listArb.head().next().next().setArb(listArb.head());
        // cout << listArb << endl;
        System.out.println("List with arbitrary pointers:\n" + listArb);

        LinkedList<Integer> listArbCopy = listArb.deepCopy(); // new LinkedList<>();
        // listArbCopy.setHead(deep_copy_arbitrary_pointer(listArb.head()));

        // Change some stuff to check if the linked lists are really separate
        listArbCopy.head().next().setValue(123);
        listArb.head().setValue(321);

        System.out.println("Updated List with arbitrary pointers:\n" + listArb);
        System.out.println("Copied List with arbitrary pointers:\n" + listArbCopy);

        // Test the insertion sort
        list.removeFront();
        list.addFront(11);
        list.addFront(82);
        list.addFront(23);
        list.addFront(29);
        
        System.out.println("Unsorted list is: " + list);
        list.sort();
        System.out.println("Sorted list is: " + list);
        
    }
}
